typed-racket-lib/typed-racket/types/generalize.rkt

```diff
@@ -28,6 +28,10 @@
         [(? (lambda (t) (subtype t -FloatComplex))) -FloatComplex]
         [(? (lambda (t) (subtype t -SingleFlonumComplex))) -SingleFlonumComplex]
         [(? (lambda (t) (subtype t -Number))) -Number]
+        [(? (lambda (t) (subtype t -Base-Regexp))) -Regexp]
```

Comment:
@samth The current implementation in TR generalises `#rx` to `-Regexp`, and `#px` to `-PRegexp`:

```
(let ([x #px"foo"]) (set! x #rx"foo") x)
error: eval:1:20: Type Checker: type mismatch;
  mutation only allowed with compatible types
   expected: PRegexp
   given: Regexp
   in: (set! x #rx"foo …

(let ([x #rx"foo"]) (set! x #px"foo") x) ;; works fine
```

Although the likelihood of this actually breaking anything is exceedingly low, changing `#px` so that it generalises to `-Regexp`  is in principle a breaking change:

```
;; works for now, fails if we make #px generalise to -Regexp
(let ([x #px"foo"]) (ann x PRegexp))
```

Changing it the other way round (so that `#rx` generalises to `-Base-Regexp`) would not break anything for now, as `-Base-Regexp` somehow magically turns into `-Regexp` anyway (I don't know what code is responsible of this). But if we change that "second generalisation", e.g. when making `Base-Regexp` available, then it would be a breaking change:

```
;; works for now, fails if we make #rx generalise to -Base-Regexp
(let ([x #rx"foo"]) (set! x #px"bar") x)
```

I'm willing to make the first breaking change, though (generalising `#px` to `-Regexp`), as it will most likely reduce surprises, and it is unlikely that there is some code relying on this in the wild. But I'd prefer confirmation before going down that road :)